SourceKitStressTester/Sources/StressTester/RequestDurationManager.swift

```diff
@@ -0,0 +1,99 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2021 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import Common
+import Foundation
+
+fileprivate extension Array where Element == Int {
+  /// Creates a logarithmic histogram. For each `key`, the `value` contains the
+  /// number of elements in this array that are smaller than `2 ^ key` but not
+  /// smaller than `2 ^ (key - 1)`.
+  var logHistogram: [Int: Int] {
+    var result: [Int: Int] = [:]
+    for value in self {
+      let log = Int(log2(Double(value))) + 1
+      result[log, default: 0] += 1
+    }
+    return result
+  }
+
+  /// The average value in this array, rounded to an integer or 0 if the array
+  /// is empty
+  var average: Int {
+    if self.isEmpty {
+      return 0
+    } else {
+      return self.reduce(0, { $0 + $1 }) / self.count
+    }
+  }
+}
+
+/// Captures aggregated information about executing a certain request kind on a
+/// certain file.
+struct AggregatedRequestDurations: Codable {
+  /// The log histogram created by the extension on `Array` above
+  let logHistogram: [Int: Int]
+
+  /// The average number of instructions a request took to execute
+  let average: Int
+
+  /// The total number of requests of a given kind executed for a given file
+  let requestsExecuted: Int
+
+  init(instructionCounts: [Int]) {
+    logHistogram = instructionCounts.logHistogram
+    average = instructionCounts.average
+    requestsExecuted = instructionCounts.count
+  }
+}
+
+/// Contains aggregated information for all request kinds of all files executed
+/// by the stress tester.
+fileprivate struct RequestDurations: Codable {
+  // FIXME: We should be using `RequestKind` as the inner key but that causes
+  // the inner dictionary to be serialized as an array (rdar://78099769)
+  /// Maps file paths to request kinds to aggregated request duration information
+  var files: [String: [String: AggregatedRequestDurations]]
```

Comment:
Not really, it’s just a note that we can replace the inner `String` with `RequestKind` once the radar is fixed.

Reply:
I didn't realise that it was important not just for the round-trip, but also for other tools reading the json. Hence why this needs to remain String (otherwise we couldn't decode in eg. python).